src/command/restore/restore.c

```diff
@@ -0,0 +1,1744 @@
+                storageRemoveP(storageLocalWrite(), pgPath, .errorOnMissing = true);
+            }
+
+            break;
+        }
+
+        case storageTypeLink:
+        {
+            const ManifestLink *manifestLink = manifestLinkFindDefault(cleanData->manifest, manifestName, NULL);
+
+            if (manifestLink != NULL)
+            {
+                if (!strEq(manifestLink->destination, info->linkDestination))
+                {
+                    LOG_DETAIL("remove link '%s' because destination changed", strPtr(pgPath));
+                    storageRemoveP(storageLocalWrite(), pgPath, .errorOnMissing = true);
+                }
+                else
+                    restoreCleanOwnership(pgPath, manifestLink->user, manifestLink->group, info->userId, info->groupId, false);
+            }
+            else
+            {
+                LOG_DETAIL("remove invalid link '%s'", strPtr(pgPath));
+                storageRemoveP(storageLocalWrite(), pgPath, .errorOnMissing = true);
+            }
+
+            break;
+        }
+
+        case storageTypePath:
+        {
+            const ManifestPath *manifestPath = manifestPathFindDefault(cleanData->manifest, manifestName, NULL);
+
+            if (manifestPath != NULL)
+            {
+                // Check ownership/permissions
+                if (dotPath)
+                {
+                    restoreCleanOwnership(pgPath, manifestPath->user, manifestPath->group, info->userId, info->groupId, false);
+                    restoreCleanMode(pgPath, manifestPath->mode, info);
+                }
+                // Recurse into the path
+                else
+                {
+                    RestoreCleanCallbackData cleanDataSub = *cleanData;
+                    cleanDataSub.targetName = strNewFmt("%s/%s", strPtr(cleanData->targetName), strPtr(info->name));
+                    cleanDataSub.targetPath = strNewFmt("%s/%s", strPtr(cleanData->targetPath), strPtr(info->name));
+                    cleanDataSub.basePath = false;
+
+                    storageInfoListP(
+                        storageLocalWrite(), cleanDataSub.targetPath, restoreCleanInfoListCallback, &cleanDataSub,
+                        .errorOnMissing = true, .sortOrder = sortOrderAsc);
+                }
+            }
+            else
+            {
+                LOG_DETAIL("remove invalid path '%s'", strPtr(pgPath));
+                storagePathRemoveP(storageLocalWrite(), pgPath, .errorOnMissing = true, .recurse = true);
+            }
+
+            break;
+        }
+
+        // Special file types cannot exist in the manifest so just delete them
+        case storageTypeSpecial:
+        {
+            LOG_DETAIL("remove special file '%s'", strPtr(pgPath));
+            storageRemoveP(storageLocalWrite(), pgPath, .errorOnMissing = true);
+            break;
+        }
+    }
+
+    FUNCTION_TEST_RETURN_VOID();
+}
+
+static void
+restoreClean(Manifest *manifest)
```

Comment:
Might be nice to call this restoreCleanBuild

Reply:
Done.